### PR TITLE
Memory backing  GSoC

### PR DESCRIPTION
**Pull request by gqlo**
### What does this PR do?
In response to feature request #58167 

Support Memory Backing which enables how virtual memory pages are backed by host pages. configurable parameters are: 
- `hugepages`
- `nosharepages`
- `locked`
- `source`
- `access` 
- `discard` 

Detailed explanation of these parameters can be found at [https://libvirt.org/formatdomain.html#memory-backing](Libvirt Domain XML)

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by gqlo** [comment]
The `ci/lint` failing test does not seem be related to the changes in `virt module`.  Will wait for @cbosdo come back from his vacation to investigate further. :)

File: salt/modules/virt.py [hunk 1]
```diff
@@ -2205,6 +2232,32 @@ def init(
     min_guarantee
         the guaranteed minimum memory allocation for the guest
 
+    hugepages
+         memory allocated using ``hugepages`` instead of the normal native page size. The value of this parameter is the
+         size of the memory page. For example ``"hugepages": "1g"``.
```
Comment (cbosdo) [inline salt/modules/virt.py:2237]: [libvirt doc](https://libvirt.org/formatdomain.html#memory-backing) mentions multiple huge pages pages and has them assigned to a cpuset. How comes this is not handled here?

File: salt/modules/virt.py [hunk 2]
```diff
@@ -3006,6 +3063,55 @@ def _set_vcpu(node, value):
...
+            "get": lambda n: str(n.get("mode")) if n.get("mode") else None,
+            "set": lambda n, v: n.set("mode", str(v)),
+        },
+        {
+            "path": "mem:allocation",
+            "xpath": "memoryBacking/allocation",
+            "get": lambda n: str(n.get("mode")) if n.get("mode") else None,
+            "set": lambda n, v: n.set("mode", str(v)),
+        },
+        {"path": "mem:discard", "xpath": "memoryBacking/discard"},
```
Comment (cbosdo) [inline salt/modules/virt.py:3114]: duplicate entry

File: salt/modules/virt.py [hunk 3]
```diff
@@ -3006,6 +3063,55 @@ def _set_vcpu(node, value):
...
+        {"path": "mem:nosharepages", "xpath": "memoryBacking/nosharepages"},
+        {"path": "mem:locked", "xpath": "memoryBacking/locked"},
```
Comment (cbosdo) [inline salt/modules/virt.py:3095]: I wonder what happens with those: they are not supposed to have text nor attributes

File: salt/utils/xmlutil.py [hunk 4]
```diff
@@ -165,6 +165,7 @@ def clean_node(parent_map, node, ignored=None):
         len(node.attrib.keys() - (ignored or [])) == 0
         and not list(node)
         and not has_text
+        and parent
```
Comment (cbosdo) [inline salt/utils/xmlutil.py:168]: You should move that hunk into a separate commit. Adding a unit test in `test_xmlutil.py` for it would be nice too.

**Review by cbosdo — CHANGES_REQUESTED** [review]
There is an issue with the hugepage parameter. See my comments inline

File: salt/modules/virt.py (hunk 1 above)
Comment (gqlo) [reply salt/modules/virt.py:2301]: new code is updated to handle assigning of ``cpuset`` :)

**Comment by gqlo** [comment]
re-run ci/pre-commit

**Comment by gqlo** [comment]
Commits of this PR is included in PR [https://github.com/saltstack/salt/pull/58196](url)
